Implement `From<Vec<T>>` and `Into<Vec<T>>` for `VecDeque<T>`

davidhewitt opened the pull request:
Closes #32848

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @alexcrichton (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

davidhewitt commented:
Ah, looks like I got the stability attribute wrong. Do I need to change the feature from rust1 to e.g. vecdeque_vec_conversions ?

apasel422 commented on src/libcollections/vec_deque.rs:
```diff
@@ -2121,6 +2122,42 @@ impl<T: fmt::Debug> fmt::Debug for VecDeque<T> {
+impl<T> From<Vec<T>> for VecDeque<T> {
+    fn from(other: Vec<T>) -> Self {
+        unsafe {
+            let other_buf: *mut T = mem::transmute(other.as_ptr());
```
If you declare `other` to be `mut`, this can just be:

```rust
let other_buf = other.as_mut_ptr();
```

(The transmute is unnecessary.)

apasel422 commented:
@davidhewitt These impls should probably start off unstable. You can use `vecdeque_vec_conversions` as the feature name, and #32848 as the issue number. I forgot that we'd need to keep a tracking issue open, so let's actually remove "Closes #32848" from the PR message.

apasel422 commented on src/libcollections/vec_deque.rs:
```diff
@@ -2121,6 +2122,42 @@ impl<T: fmt::Debug> fmt::Debug for VecDeque<T> {
+
+            // Do this to force resize to correct size;
+            // otherwise we may get not_power_of_two assert failures!
+            out.reserve(0);
```
It'd be good to test for the power-of-two invariant in your test for this method.

apasel422 commented on src/libcollections/vec_deque.rs:
```diff
@@ -2401,4 +2438,24 @@ mod tests {
     }
+
+    #[test]
+    fn test_from_vec() {
```
Can you add some additional test cases here with varying vec lengths and capacities? It might be easiest to do something like:

```rust
#[test]
fn test_from_vec() {
    for cap in 0..35 {
        for len in 0..cap + 1 {
            let mut vec = Vec::with_capacity(cap);
            vec.extend(0..len);

            let vd = VecDeque::from(vec.clone());
            assert!(vd.cap().is_power_of_two());
            assert_eq!(vd.len(), vec.len());
            assert!(vd.into_iter().eq(vec));
        }
    }
}
```

apasel422 commented on src/libcollections/vec_deque.rs:
```diff
@@ -2401,4 +2438,24 @@ mod tests {
+    }
+
+    #[test]
+    fn test_into_vec() {
```
This could also benefit from some other test cases to ensure that the use of `wrap_copy` is correct, with varying item layouts (contiguous, wrapping, etc).

alexcrichton commented:
Thanks for the PR @davidhewitt! @apasel422 unfortunately we don't actually read stability annotations on trait impls right now, so these'll be insta-stable if they land (so `#[stable]` is fine).

It's also interesting implementing the `Into` trait instead of the `From` trait as `From<VecDeque<T>> for Vec<T>`. Right now we pretty consistently implement `From` rather than `Into` (as it also gives you the `into` method). Would that be possible to do, or is that not allowed because of privacy just yet?

apasel422 commented:
@alexcrichton Hmm, I didn't even consider the `From`/`Into` distinction when I filed the issue, but yes, we should use whatever convention we've been going with.

davidhewitt commented:
Hi guys, I'm working on some corrections to this PR, might be a couple of evenings before they land if that's ok?

R.E. the From / Into - shall I rework the Into into a From on the Vec class?

alexcrichton commented:
@davidhewitt no worries! Feel free to ping this PR when it's updated. And yeah let's try to put `From` on `Vec` directly, but if it's too hard we can discuss that at libs triage.

davidhewitt commented:
@apasel422 @alexcrichton So I've taken the first round of comments on board and rewritten this patch with better tests, and covered some corner cases uncovered during testing. Thanks for the feedback, I feel I learned some good stuff with how to make better tests there!

RE. putting the `From` on `Vec`, I can do this as long as it's implemented in vec_deque.rs, and not vec.rs. That feels a little weird to me, but privacy stops it working the other way around. Does that feel okay with you guys?

alexcrichton commented on src/libcollections/vec_deque.rs:
```diff
@@ -2121,6 +2122,100 @@ impl<T: fmt::Debug> fmt::Debug for VecDeque<T> {
+                        for i in left_edge..right_edge {
+                            right_offset = (i - left_edge) % (cap - right_edge);
+                            let src: isize = (right_edge + right_offset) as isize;
+                            println!("swap {} to {}; cap {}; left {}; len: {}; right {};", src, i, cap, left_edge, len, right_edge);
```
Looks like a stray println?

alexcrichton commented:
Thanks for the update @davidhewitt! The libs team discussed this in triage yesterday and the conclusion was that this is good-to-go (especially now that both impls are `From` rather than one of each).

I'll leave the r+ to @apasel422 though :)

@bors: delegate=apasel422

bors commented:
:v: @apasel422 can now approve this pull request

davidhewitt commented:
Thanks @alexcrichton!

@apasel422 - I took "Closes #32848" out of the PR message, just checking that's what you wanted?

apasel422 commented on src/libcollections/vec_deque.rs:
```diff
@@ -2121,6 +2122,99 @@ impl<T: fmt::Debug> fmt::Debug for VecDeque<T> {
+            let cap = other.cap();
+
+            // Need to move the ring to the front of the buffer, as vec will expect this.
+            if other.is_contiguous() {
```
Would it be possible to ensure that `test_vec_from_vecdeque` explicitly tests each of the four cases?

apasel422 commented:
This looks good, pending some additional tests. Once you've done that, could you squash all the commits into one?

davidhewitt commented:
I rewrote test_vec_from_vecdeque to expand out the loops, to label which configurations of the loop variables will be testing which bits of the ring-straightening algorithm. Also squashed the commits.

Is there other tests you have in mind?

apasel422 commented:
Looks like there are a few lines longer than 100 characters. Otherwise, this is ready to merge.

apasel422 commented:
@davidhewitt Looks like this failed with a syntax error.

davidhewitt commented:
Sorry, made a mistake breaking up the lines.

apasel422 commented:
@alexcrichton Is `1.9.0` the right value for `since`?

alexcrichton commented:
Ah nah we're on to 1.10.0 at this point. Should write a tidy check for that somehow...

davidhewitt commented:
Ah okay - I'll fix that up tonight!

apasel422 commented:
@bors r+ 1861951

bors commented:
:hourglass: Testing commit 1861951 with merge 14f61c8...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/8816), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/8851), [auto-linux-64-cargotest](http://buildbot.rust-lang.org/builders/auto-linux-64-cargotest/builds/140), [auto-linux-64-debug-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/2084), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/8790), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/8779), [auto-linux-64-opt-mir](http://buildbot.rust-lang.org/builders/auto-linux-64-opt-mir/builds/355), [auto-linux-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-linux-64-opt-rustbuild/builds/764), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/8773), [auto-linux-64-x-armhf](http://buildbot.rust-lang.org/builders/auto-linux-64-x-armhf/builds/359), [auto-linux-64-x-armsf](http://buildbot.rust-lang.org/builders/auto-linux-64-x-armsf/builds/354), [auto-linux-64-x-freebsd](http://buildbot.rust-lang.org/builders/auto-linux-64-x-freebsd/builds/364), [auto-linux-64-x-netbsd](http://buildbot.rust-lang.org/builders/auto-linux-64-x-netbsd/builds/353), [auto-linux-cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/1993), [auto-linux-musl-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/3934), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/8824), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/8833), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/8796), [auto-mac-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-mac-64-opt-rustbuild/builds/757), [auto-mac-ios-opt](http://buildbot.rust-lang.org/builders/auto-mac-ios-opt/builds/951), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/3848), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/3863), [auto-win-gnu-32-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt-rustbuild/builds/765), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/3859), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/3837), [auto-win-msvc-32-cross-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-cross-opt/builds/232), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/3057), [auto-win-msvc-64-cargotest](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-cargotest/builds/140), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/3848), [auto-win-msvc-64-opt-mir](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-mir/builds/353), [auto-win-msvc-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-rustbuild/builds/771)
